### PR TITLE
update integration tests to work against cyrus imapd

### PR DESCRIPTION
**Pull request by urkle**
Adds cyrus-imapd support for testing.

---
This change is [Reviewable](https://reviewable.io/reviews/jonhoo/rust-imap/228)

### CODE REVIEW

**Comment by jonhoo** [comment]
Thanks! Two initial observations:

1. This seems to _replace_ Greenmail with Cyrus, which I don't think is what we should do. Instead, I'd prefer to test against _both_. I'm fine if we only run a subset of CI against both though. Coverage for example we can just run against one.
2. Is it feasible for us to run Cyrus through `services` the same way we do with Greenmail rather than constructing the Docker image inline with the build steps?

**Comment by urkle** [comment]
> Thanks! Two initial observations:
> 
>     1. This seems to _replace_ Greenmail with Cyrus, which I don't think is what we should do. Instead, I'd prefer to test against _both_. I'm fine if we only run a subset of CI against both though. Coverage for example we can just run against one.

Yeah, I can do that. (and sounds very reasonable)   What I'll do to make things easier for controlling what tests are run is move the ACL tests that I'll write to a separate integration test file entirely. 

>     2. Is it feasible for us to run Cyrus through `services` the same way we do with Greenmail rather than constructing the Docker image inline with the build steps?

In order to do that I'd have to publish the pre-built containers to docker hub.  and once I get all the issues sorted (after writing the ACL tests) then I'll do that and that'll make this tests run much faster as well.

**Comment by urkle** [comment]
@jonhoo I have reworked this PR to now only add a test-run for cyrus-imapd and pushed up an single image to my docker hub that spins up a service somewhat mimicking greenmail.

File: tests/imap_integration.rs [hunk 1]
```diff
@@ -145,12 +196,11 @@ fn inbox() {
         .any(|m| m == &imap::types::UnsolicitedResponse::Recent(2)));
 
     // let's see that we can also fetch the e-mails
-    let fetch = c.fetch("1", "(ALL UID)").unwrap();
+    let fetch = c.fetch("1", "(ENVELOPE INTERNALDATE UID)").unwrap();
```
Comment (jonhoo) [inline tests/imap_integration.rs:199]: Why all these changes away from `ALL`?

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -32,3 +32,27 @@ jobs:
...
+          profile: minimal
+          toolchain: ${{ matrix.toolchain }}
+      - uses: actions/checkout@v2
+      - name: cargo test
+        uses: actions-rs/cargo@v1
+        with:
+          command: test
+    services:
+      cyrus_imapd:
+        image: outoforder/cyrus-imapd-tester:latest
```
Comment (jonhoo) [inline .github/workflows/test.yml:53]: One worry I have with this is that version of Cyrus is likely to get out of date over time unless you commit to updating it. Is there not a standard Cyrus Docker image that's regularly maintained we can use?

File: .github/workflows/test.yml [hunk 3]
```diff
@@ -6,7 +6,7 @@ name: cargo test
 jobs:
   test:
     runs-on: ubuntu-latest
-    name: ${{ matrix.toolchain }}
+    name: Run tests with greenmail ${{ matrix.toolchain }}
```
Comment (jonhoo) [inline .github/workflows/test.yml:9]:
```suggestion
    name: greenmail/${{ matrix.toolchain }}
```

File: .github/workflows/test.yml [hunk 4]
```diff
@@ -32,3 +32,27 @@ jobs:
           - 3995:3995
         env:
           GREENMAIL_OPTS: "-Dgreenmail.setup.test.all -Dgreenmail.hostname=0.0.0.0 -Dgreenmail.auth.disabled -Dgreenmail.verbose"
+  test_cyrus:
+    runs-on: ubuntu-latest
+    name: Run tests with cyrus ${{ matrix.toolchain }}
```
Comment (jonhoo) [inline .github/workflows/test.yml:37]:
```suggestion
    name: cyrus/${{ matrix.toolchain }}
```

**Review by jonhoo — CHANGES_REQUESTED** [review]
Oh, yeah, this looks much nicer! Left another few notes.

File: tests/imap_integration.rs (hunk 1 above)
Comment (urkle) [reply tests/imap_integration.rs:199]: Cyrus IMAP doesn't support ALL

File: .github/workflows/test.yml (hunk 2 above)
Comment (urkle) [reply .github/workflows/test.yml:53]: No, there is no official cyrus docker image at all.  (I had looked for that first when setting up testing for the app I'm building).   But this docker image is based on Rocky8 (derived from RHEL8) and thus that base image will be maintained for quite a while.  And I'll setup a github repo here on the next few days with the docker file.   And in reality the supported features of cyrus for IMAP haven't changed much in years.  (e.g. this is version 3.0.7 and my other servers are running 2.4.x and the functionality is the same between them). 

But as I'm maintaining several mail servers (on Cent 6, and moving toward Rocky 8 or Rocky 9) and writing admin software that is utilizing your crate, I'll be maintaining this for a while.

What I'll do next is setup a github repo with the docker image(s) and set it up so it'll pop out a Rocky 8 and Rocky 9 base image with the same setup.

File: tests/imap_integration.rs (hunk 1 above)
Comment (jonhoo) [reply tests/imap_integration.rs:199]: That's super weird — it's in the base IMAP RFC 3501 after all! For `FETCH` it is [defined as equivalent](https://datatracker.ietf.org/doc/html/rfc3501#page-55) to
```
FLAGS INTERNALDATE RFC822.SIZE ENVELOPE
```
so maybe we should use that?

File: tests/imap_integration.rs [hunk 5]
```diff
@@ -19,9 +19,61 @@ fn tls() -> native_tls::TlsConnector {
...
+fn test_smtps_port() -> u16 {
+    std::env::var("TEST_SMTPS_PORT")
+        .unwrap_or("3465".to_string())
+        .parse()
+        .unwrap_or(3465)
+}
+
+fn clean_mailbox(session: &mut imap::Session<native_tls::TlsStream<TcpStream>>) {
+    session.select("INBOX").unwrap();
+    let inbox = session.search("ALL").unwrap();
```
Comment (jonhoo) [inline tests/imap_integration.rs:54]: Hah, so Cyrus supports `ALL` for `SEARCH`, just _not_ for `FETCH`. Very weird.

**Review by jonhoo — APPROVED** [review]

**Comment by jonhoo** [comment]
Huh, it looks like `connect_insecure_then_secure` is suddenly failing with Greenmail?

File: tests/imap_integration.rs (hunk 1 above)
Comment (urkle) [reply tests/imap_integration.rs:199]: I saw that as well that it was int he 3501 RFC, but not the earlier 2060 spec. both of which claim to  4rev1.

I changed it to be what was actually be checked in the tests vs everything.

File: tests/imap_integration.rs (hunk 5 above)
Comment (urkle) [reply tests/imap_integration.rs:54]: yes.. very strange.

**Comment by urkle** [comment]
> 

I just saw the failed tests come back.. I'll look into it and see what got broke and fix.

**Comment by urkle** [comment]
@jonhoo ok.. so I had removed the `#[ignore]` on the starttls test as cyrus does support STARTTLS.  I pushed a fixup that re-added the `#[ignore]` and adjusted the cyrus test to run the ignored tests (for now).

I think the easier approach going forward would be to move the integration tests that cannot be run on greenmail to a separate integration test script.  Then we setup the test to run like this in the CI setup
```
cargo test --lib
cargo test --examples
cargo test --doc
cargo test --test imap_integration
cargo test --test extra_integration # for cyrus only.
```

Unfortunately the cargo test runner does not allow you to run all non-integration tests in one go. (for some reason --doc can't be used with other arguments)

File: tests/imap_integration.rs [hunk 6]
```diff
@@ -288,7 +343,7 @@ fn list() {
     let mut s = session("readonly-test@localhost");
     s.select("INBOX").unwrap();
     let subdirs = s.list(None, Some("%")).unwrap();
-    assert_eq!(subdirs.len(), 0);
+    assert_eq!(subdirs.len(), 1);
```
Comment (jonhoo) [inline tests/imap_integration.rs:346]: This feels like a weird change. Why did the length suddenly change?

**Review by jonhoo — COMMENTED** [review]
Another way to do this is to have test features. We could either have a feature like `test_on_cyrus`, or we could make it more general along the lines of `test_with_all_capabilities`.

File: tests/imap_integration.rs (hunk 6 above)
Comment (jonhoo) [reply tests/imap_integration.rs:346]: Oh, right, I guess the test was ignored before so it was probably wrong.

Comment (urkle) [reply tests/imap_integration.rs:346]: :);.  I only changed it because I currently have cyrus running all ignored tests.

**Comment by urkle** [comment]
> Another way to do this is to have test features. We could either have a feature like `test_on_cyrus`, or we could make it more general along the lines of `test_with_all_capabilities`.

Ahh.. using the cargo feature stuff:).  that is an option.   Once we land this we can play w/ that over on the other 2 branches.

**Comment by jonhoo** [comment]
:tada:

**Pull request merged**